### PR TITLE
Cleanup and improve test coverage for Concat tests

### PR DESCRIPTION
**Pull request by jamesqo**
These changes convert most of the `Concat` tests to use `[Theory]` and improve test coverage, since the asserts are more centralized in one place. I added tests for the relevant codepaths mentioned in #11492; the case where we concat two collections, then append an enumerable is handled indirectly by `ChainedCollectionConcatData`, since an `Enumerable.Empty<T>().Concat(x.ToList())` gets passed to the theory and that is eventually compared to `.Concat(ForceNotCollection(Array.Empty<T>()))`.

Separated out from #11492 since I felt the PR was already a little long.

cc @stephentoub

### CODE REVIEW

**Comment by stephentoub** [comment]
cc: @VSadov, @JonHanna

File: src/System.Linq/tests/ConcatTests.cs [hunk 1]
```diff
@@ -4,68 +4,48 @@
...
-        [Fact]
-        public void BothEmpty()
+        [Theory]
+        [InlineData(new int[] { 2, 3, 2, 4, 5 }, new int[] { 1, 9, 4 })]
+        public void SameResultsWithQueryAndRepeatCalls(IEnumerable<int> first, IEnumerable<int> second)
         {
-            int[] first = { };
-            int[] second = { };
-            Assert.Empty(first.Concat(second));
+            // workaround: xUnit type inference doesn't work if the input type is not T (like IEnumerable<T>)
```
Comment (hughbe) [inline src/System.Linq/tests/ConcatTests.cs]: I made PR for this several days ago
Take a look if you'd like, https://github.com/xunit/xunit/pull/965

File: src/System.Linq/tests/ConcatTests.cs [hunk 2]
```diff
@@ -4,68 +4,48 @@
...
-        public void SameResultsRepeatCallsStringQuery()
-        {
-            var q1 = from x1 in new[] { "AAA", String.Empty, "q", "C", "#", "!@#$%^", "0987654321", "Calling Twice" }
-                     select x1;
-            var q2 = from x2 in new[] { "!@#$%^", "C", "AAA", "", "Calling Twice", "SoS" }
-                     select x2;
-
-            Assert.Equal(q1.Concat(q2), q1.Concat(q2));
-        }
+        private delegate IEnumerable<T> EnumerableTransform<T>(IEnumerable<T> input);
```
Comment (stephentoub) [inline src/System.Linq/tests/ConcatTests.cs]: Why a custom delegate rather than just using ```Func<IEnumerable<T>, IEnumerable<T>>```?

File: src/System.Linq/tests/ConcatTests.cs [hunk 3]
```diff
@@ -4,68 +4,48 @@
...
 namespace System.Linq.Tests
 {
     public class ConcatTests : EnumerableTests
     {
-        [Fact]
-        public void SameResultsRepeatCallsIntQuery()
-        {
-            var q1 = from x1 in new int?[] { 2, 3, null, 2, null, 4, 5 }
-                     select x1;
-            var q2 = from x2 in new int?[] { 1, 9, null, 4 }
```
Comment (stephentoub) [inline src/System.Linq/tests/ConcatTests.cs]: Why did you get rid of the nullables?

File: src/System.Linq/tests/ConcatTests.cs [hunk 4]
```diff
@@ -4,68 +4,48 @@
...
-        public void NonEmptyAndNonEmpty()
+        [Theory]
+        [InlineData(new int[] { }, new int[] { }, new int[] { })] // Both inputs are empty
+        [InlineData(new int[] { }, new int[] { 2, 6, 4, 6, 2 }, new int[] { 2, 6, 4, 6, 2 })] // One is empty
+        [InlineData(new int[] { 2, 3, 5, 9 }, new int[] { 8, 10 }, new int[] { 2, 3, 5, 9, 8, 10 })] // Neither side is empty
+        public void PossiblyEmptyInputs(IEnumerable<int> first, IEnumerable<int> second, IEnumerable<int> expected)
         {
-            int?[] first = { 2, null, 3, 5, 9 };
-            int?[] second = { null, 8, 10 };
-            int?[] expected = { 2, null, 3, 5, 9, null, 8, 10 };
```
Comment (stephentoub) [inline src/System.Linq/tests/ConcatTests.cs]: Similarly, why did you remove the nulls?

File: src/System.Linq/tests/ConcatTests.cs [hunk 5]
```diff
@@ -89,142 +70,138 @@ public void SecondNull()
...
-        [Fact]
-        public void TwoArraySources()
+        private static void VerifyEqualsWorker<T>(IEnumerable<T> expected, IEnumerable<T> actual)
         {
-            VerifyEquals(
-                Enumerable.Range(0, 7),
-                Enumerable.Range(0, 3).ToArray().Concat(Enumerable.Range(3, 4).ToArray()));
+            // All of these transforms should take an enumerable and produce
+            // another enumerable with the same contents.
+            var transforms = new List<EnumerableTransform<T>>
```
Comment (stephentoub) [inline src/System.Linq/tests/ConcatTests.cs]: Nit: ```transforms``` suggests to me as a reader that the values are being changed.  I'd suggest at least using the word ```identity``` in it somehow, e.g. ```identityTransforms```.

File: src/System.Linq/tests/ConcatTests.cs [hunk 6]
```diff
@@ -89,142 +70,138 @@ public void SecondNull()
...
+                e => e,
+                e => e.ToArray(),
+                e => e.ToList(),
+                e => e.Select(i => i),
+                e => e.Select(i => i).ToArray(),
+                e => e.Where(i => true),
+                e => e.Concat(Array.Empty<T>()),
+                e => ForceNotCollection(e),
+                e => e.Concat(ForceNotCollection(Array.Empty<T>()))
+            };
```
Comment (stephentoub) [inline src/System.Linq/tests/ConcatTests.cs]: Why are all of these necessary?  Do we have special code paths in the src for, for example, Select or Where on the result of a Concat?

File: src/System.Linq/tests/ConcatTests.cs [hunk 7]
```diff
@@ -89,142 +70,138 @@ public void SecondNull()
...
+            // of expected against all transforms of actual.
+            foreach (var outTransform in transforms)
+            {
+                foreach (var inTransform in transforms)
+                {
+                    IEnumerable<T> sameExpected = outTransform(expected);
+                    IEnumerable<T> sameActual = inTransform(actual);
+
+                    Assert.Equal(sameExpected, sameActual);
+                    Assert.Equal(sameExpected.Count(), sameActual.Count());
```
Comment (stephentoub) [inline src/System.Linq/tests/ConcatTests.cs]: If sameExpected equals sameActual, as is checked in the previous assert, how could this subsequent assert for their counts ever fail?

File: src/System.Linq/tests/ConcatTests.cs [hunk 8]
```diff
@@ -89,142 +70,138 @@ public void SecondNull()
...
+        public static IEnumerable<object[]> ArraySourcesData() => GenerateSourcesData(outerTransform: e => e.ToArray());
 
-        [Fact]
-        public void TwoNonCollectionSources()
-        {
-            VerifyEquals(
-                Enumerable.Range(0, 7),
-                NumberRangeGuaranteedNotCollectionType(0, 3).Concat(NumberRangeGuaranteedNotCollectionType(3, 4).ToArray().Select(i => i)));
-        }
+        public static IEnumerable<object[]> ArraySelectSourcesData() => GenerateSourcesData(outerTransform: e => e.Select(i => i).ToArray());
```
Comment (stephentoub) [inline src/System.Linq/tests/ConcatTests.cs]: Shouldn't this be SelectArray rather than ArraySelect?

File: src/System.Linq/tests/ConcatTests.cs [hunk 9]
```diff
@@ -89,142 +70,138 @@ public void SecondNull()
...
-            Assert.Equal(Enumerable.Empty<int>(), source);
+
+            Assert.Equal(toConcat.Count() * times, source.Count());
+            VerifyEqualsWorker(expected, source);
+        }
+
+        public static IEnumerable<object[]> ManyConcatsData()
+        {
+            yield return new object[] { Enumerable.Empty<int>(), 256, Enumerable.Empty<int>() };
+            yield return new object[] { Enumerable.Repeat(6, 1), 256, Enumerable.Repeat(6, 256) };
```
Comment (stephentoub) [inline src/System.Linq/tests/ConcatTests.cs]: Should we have an input here where toConcat is a collection rather than just an enumerable?

Comment (stephentoub) [inline src/System.Linq/tests/ConcatTests.cs]: Also, by using the same value over and over, your change loses some of the verification in the original, e.g. if the concat messed up the ordering, the previous version of the test would catch it, but your change wouldn't.

File: src/System.Linq/tests/ConcatTests.cs (hunk 2 above)
Comment (jamesqo) [inline src/System.Linq/tests/ConcatTests.cs]: I found this more readable-- having a `List<Func<IEnumerable<T>, IEnumerable<T>>>` seems quite verbose.

File: src/System.Linq/tests/ConcatTests.cs (hunk 3 above)
Comment (jamesqo) [inline src/System.Linq/tests/ConcatTests.cs]: `InlineData` does not work with `Nullable<int>` since the compiler says it's not a constant expression, so I had to change it to a plain int.

File: src/System.Linq/tests/ConcatTests.cs (hunk 2 above)
Comment (stephentoub) [inline src/System.Linq/tests/ConcatTests.cs]: FWIW, I find it less readable.  At the usage site I don't know what an EnumerableTransform is, making it less obvious what's happening with type inference in the lambdas.

File: src/System.Linq/tests/ConcatTests.cs (hunk 6 above)
Comment (jamesqo) [inline src/System.Linq/tests/ConcatTests.cs]: I only kept them since we were testing `Where(i => true)` and `Select(i => i)` before this change, and I didn't want to accidentally remove code coverage. I think we can get rid of the `Where` since that's handled by `ForceNotCollection`, and `Select(i => i).ToArray()` should be the same thing as `ToArray()`. I'm keeping the `Select` part though, since that's the only `IIListProvider` that is not a List.

File: src/System.Linq/tests/ConcatTests.cs (hunk 7 above)
Comment (jamesqo) [inline src/System.Linq/tests/ConcatTests.cs]: You're correct-- I only included this assert because the previous version of `VerifyEquals` had it, but I see now that it's redundant. I'll remove this.

File: src/System.Linq/tests/ConcatTests.cs (hunk 9 above)
Comment (jamesqo) [inline src/System.Linq/tests/ConcatTests.cs]: > Should we have an input here where toConcat is a collection rather than just an enumerable?

Maybe this should be done in the theory to avoid code repetition--

```cs
foreach (var equivalentEnumerable in new[] { toConcat, toConcat.ToList(), ForceNotCollection(toConcat) ... })
```

**Comment by jamesqo** [comment]
@stephentoub Thanks for the PR feedback, I think I've implemented all of the changes you've suggested.

**Comment by stephentoub** [comment]
Thanks, @jamesqo. LGTM.
